sol/main.c

```diff
@@ -600,24 +600,41 @@ alloc_sol_conf(void)
 }
 
 int
-gk_solicitor_enqueue(struct sol_config *sol_conf, struct rte_mbuf *pkt,
-	uint8_t priority)
+gk_solicitor_enqueue_bulk(struct sol_config *sol_conf,
+	struct rte_mbuf **pkts, uint8_t *priorities, uint16_t num_pkts)
 {
-	struct priority_req *req_node;
+	int i;
+	int tot_bytes = 0;
+	struct priority_req *req_nodes[num_pkts];
 
-	if (priority > GK_MAX_REQ_PRIORITY) {
-		SOL_LOG(ERR, "Trying to enqueue a request with priority %hhu, but should be in range [0, %d]\n",
-			priority, GK_MAX_REQ_PRIORITY);
+	if (unlikely(rte_mempool_get_bulk(sol_conf->mb.pool,
+			(void **)req_nodes, num_pkts) != 0))
 		return -1;
+
+	for (i = 0; i < num_pkts; i++) {
+		if (priorities[i] > GK_MAX_REQ_PRIORITY) {
+			SOL_LOG(ERR, "Trying to enqueue a request with priority %hhu, but should be in range [0, %d]\n",
+				priorities[i], GK_MAX_REQ_PRIORITY);
+			rte_mempool_put_bulk(sol_conf->mb.pool,
+				(void **)req_nodes, num_pkts);
+			return -1;
+		}
+
+		INIT_LIST_HEAD(&req_nodes[i]->list);
+		req_nodes[i]->pkt = pkts[i];
+		req_nodes[i]->priority = priorities[i];
+
+		tot_bytes += rte_pktmbuf_pkt_len(pkts[i]);
 	}
 
-	req_node = mb_alloc_entry(&sol_conf->mb);
-	if (req_node == NULL)
+	if (unlikely(rte_ring_mp_enqueue_bulk(sol_conf->mb.ring,
+			(void **)req_nodes, num_pkts, NULL) != num_pkts)) {
+		SOL_LOG(ERR, "Trying to enqueue a bulk of %hu requests\n",
+			num_pkts);
+		rte_mempool_put_bulk(sol_conf->mb.pool,
+			(void **)req_nodes, num_pkts);
```

Comment:
According to the [DPDK documentation](https://doc.dpdk.org/api/rte__ring_8h.html#adb91b9286b8fa4b962c8e01accd089b9), ``rte_ring_mp_enqueue_bulk()`` will return either ``0`` or ``n``. So, I think we can release all the ``req_nodes`` here.

Reply:
I've missed that in the documentation; thank you. But given the code interface, we should write code like the code we write for `rte_eth_tx_burst()` to avoid having a bug when DPDK changes the semantic of the interface. We've seen them changing error codes without even updating the documentation. Thus, we should assume that `rte_ring_mp_enqueue_bulk()` might return anything in between `0` and `n` inclusive those values.

Reply:
Good point! Will do.